src/main/tools/build-runfiles-windows.cc

```diff
@@ -12,16 +12,389 @@
+      return true;
+    }
+
+    // If we couldn't create symlink, print out an error message and exit.
+    if (GetLastError() == ERROR_PRIVILEGE_NOT_HELD) {
+      die(L"CreateSymbolicLinkW failed:\n%hs\n",
+          "Bazel needs to create symlink for building runfiles tree.\n"
+          "Creating symlink on Windows requires either of the following:\n"
+          "    1. Program is running with elevated privileges (Admin rights).\n"
+          "    2. The system version is Windows 10 Creators Update (1703) or later and "
+          "developer mode is enabled.",
+          GetLastErrorString().c_str());
+    } else {
+      die(L"CreateSymbolicLinkW failed: %hs", GetLastErrorString().c_str());
+    }
+
+    return true;
+  }
+
+  // This function scan the current directory, remove all files/symlinks/directories that
+  // are not presented in manifest file.
+  // If a symlink already exists and points to the correct target, this function erases
+  // its entry from manifest_file_map, so that we won't recreate it.
+  void ScanTreeAndPrune(const wstring& path) {
+    static const wstring kDot(L".");
+    static const wstring kDotDot(L"..");
+
+    WIN32_FIND_DATAW metadata;
+    HANDLE handle = ::FindFirstFileW((path + L"\\*").c_str(), &metadata);
+    if (handle == INVALID_HANDLE_VALUE) {
+      return;  // directory does not exist or is empty
+    }
+
+    do {
+      if (kDot != metadata.cFileName && kDotDot != metadata.cFileName) {
+        wstring subpath = path + L"\\" + metadata.cFileName;
+        bool is_dir =
+            (metadata.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0;
+        bool is_symlink =
+            (metadata.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) != 0;
+        if (is_symlink) {
+          wstring target;
+          if (!blaze_util::ReadSymlinkW(subpath, &target)) {
+            die(L"ReadSymlinkW failed (%s): %hs",
+                subpath.c_str(), GetLastErrorString().c_str());
+          };
+
+          subpath = AsAbsoluteWindowsPath(subpath.c_str());
+          target = AsAbsoluteWindowsPath(target.c_str());
+          ManifestFileMap::iterator expected_target =
+              manifest_file_map.find(subpath);
+
+          if (expected_target == manifest_file_map.end()
+              || expected_target->second.empty()
+              || target != AsAbsoluteWindowsPath(expected_target->second.c_str())
+              || blaze_util::IsDirectoryW(target) != is_dir) {
+            if (is_dir) {
+              RemoveDirectoryOrDie(subpath);
+            } else {
+              DeleteFileOrDie(subpath);
+            }
+          } else {
+            manifest_file_map.erase(expected_target);
+          }
+        } else {
+          if (is_dir) {
+            ScanTreeAndPrune(subpath);
+            // If the directory is empty, then we remove the directory.
+            // Otherwise RemoveDirectory will fail with ERROR_DIR_NOT_EMPTY,
+            // which we can just ignore.
```

Comment:
Why can we ignore it?

Reply:
Thanks! Could you add this explanation to the code?

Reply:
Done!